packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Models/Types/SystemModelSerializationTypeProvider.cs

```diff
@@ -0,0 +1,231 @@
+    /// </summary>
+    internal sealed class SystemModelSerializationTypeProvider : TypeProvider
+    {
+        internal static readonly Parameter SerializationOptionsParameter =
+            new("options", null, typeof(ModelReaderWriterOptions), null, ValidationType.None, null);
+        private readonly CSharpType _iJsonModelTInterface;
+        private readonly CSharpType? _iJsonModelObjectInterface;
+        private readonly CSharpType _iPersistableModelTInterface;
+        private readonly CSharpType? _iPersistableModelObjectInterface;
+
+        public SystemModelSerializationTypeProvider(ModelTypeProvider model) : base(null)
+        {
+            Model = model;
+            Name = model.Name;
+            Json = BuildJsonSerialization();
+            // Initialize the serialization interfaces
+            _iJsonModelTInterface = new CSharpType(typeof(IJsonModel<>), Model.Type);
+            _iJsonModelObjectInterface = Model.IsStruct ? (CSharpType)typeof(IJsonModel<object>) : null;
+            _iPersistableModelTInterface = new CSharpType(typeof(IPersistableModel<>), Model.Type);
+            _iPersistableModelObjectInterface = Model.IsStruct ? (CSharpType)typeof(IPersistableModel<object>) : null;
+        }
+
+        public override string Name { get; }
+        public ModelTypeProvider Model { get; }
+        public JsonObjectSerialization? Json { get; }
+
+        /// <summary>
+        /// Builds the serialization methods for the model. If the serialization supports JSON, it will build the JSON serialization methods.
+        /// </summary>
+        /// <returns>A list of serialization and deserialization methods for the model.</returns>
+        protected override CSharpMethod[] BuildMethods()
+        {
+            var methods = new List<CSharpMethod>();
+
+            // Build the JSON serialization methods
+            if (Json != null)
+            {
+                foreach (var method in BuildJsonSerializationMethods())
+                {
+                    methods.Add(method);
+                }
+
+                // TO-DO: Add deserialization methods if applicable https://github.com/microsoft/typespec/issues/3330
+            }
+
+            // Build the IModel methods
+            var iModelMethods = BuildIModelMethods();
+            methods.AddRange(iModelMethods);
+
+            return methods.ToArray();
+        }
+
+        /// <summary>
+        /// Builds the types that the model type serialization implements.
+        /// </summary>
+        /// <returns>An array of <see cref="CSharpType"/> types that the model implements.</returns>
+        protected override CSharpType[] BuildImplements()
+        {
+            var interfaces = new List<CSharpType>
+            {
+                _iJsonModelTInterface,
+            };
+
+            if (_iJsonModelObjectInterface != null)
+            {
+                interfaces.Add(_iJsonModelObjectInterface);
+            }
+
+            return interfaces.ToArray();
```

Comment:
same comment on knowing the size of the array.

This occurs in a few places not marking all of them